src/lseq.rs

```diff
@@ -0,0 +1,269 @@
+//! generation, and ensures that the result is always between the two provided bounds.
+//!
+//! LSEQ is a CmRDT, to guarantee convergence it must see every operation. It also requires that
+//! they are delivered in a _causal_ order. Every deletion _must_ be applied _after_ it's
+//! corresponding insertion. To guarantee this property, use a causality barrier.
+//!
+//! [1] B. Nédelec, P. Molli, A. Mostefaoui, and E. Desmontils,
+//! “LSEQ: an adaptive structure for sequences in distributed collaborative editing,”
+//! in Proceedings of the 2013 ACM symposium on Document engineering - DocEng ’13,
+//! Florence, Italy, 2013, p. 37, doi: 10.1145/2494266.2494278.
+//!
+//! [2] S. Weiss, P. Urso, and P. Molli,
+//! “Logoot: A Scalable Optimistic Replication Algorithm for Collaborative Editing on P2P Networks,”
+//! in 2009 29th IEEE International Conference on Distributed Computing Systems,
+//! Montreal, Quebec, Canada, Jun. 2009, pp. 404–412, doi: 10.1109/ICDCS.2009.75.
+
+/// Contains the implementation of the exponential tree for LSeq
+pub mod ident;
+
+use ident::*;
+use serde::{Deserialize, Serialize};
+
+use crate::traits::CmRDT;
+
+use crate::vclock::{Actor, Dot};
+
+// SiteId can be generalized to any A if there is a way to generate a single invalid actor id at every site
+// Currently we rely on every site using the Id 0 for that purpose.
+
+/// Actor type for LSEQ sites.
+#[derive(Debug, Hash, Clone, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
+pub struct SiteId(u32);
+
+impl SiteId {
+    /// Create a `SiteId` from a `u32`.
+    pub fn new(id: u32) -> Self {
+        SiteId(id)
+    }
+}
+
+impl Default for SiteId {
+    fn default() -> Self {
+        SiteId(0)
+    }
+}
+
+/// An `Entry` to the LSEQ consists of:
+#[derive(Debug, Clone)]
+pub struct Entry<T, A: Actor> {
+    /// The identifier of the entry.
+    pub id: Identifier<A>,
+    /// The site id of the entry.
+    pub dot: Dot<A>,
+    /// The element for the entry.
+    pub c: T,
+}
+
+/// As described in the module documentation:
+///
+/// An LSEQ tree is a CRDT for storing sequences of data (Strings, ordered lists).
+/// It provides an efficient view of the stored sequence, with fast index, insertion and deletion
+/// operations.
+pub struct LSeq<T, A: Actor> {
+    seq: Vec<Entry<T, A>>,
+    gen: IdentGen<A>,
+    dot: Dot<A>,
+}
+
+/// Operations that can be performed on an LSeq tree
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(tag = "op")]
+pub enum Op<T, A: Actor> {
+    /// Insert an element
+    Insert {
+        /// Identifier to insert at
+        #[serde(flatten)]
+        id: Identifier<A>,
+        /// clock of site that issued insertion
+        dot: Dot<A>,
+        /// Element to insert
+        c: T,
+    },
+    /// Delete an element
+    Delete {
+        /// The original clock information of the insertion we're removing
+        remote: Dot<A>,
+        #[serde(flatten)]
+        /// Identifier to remove
```

Comment:
clearer if you can move the `#[serde(flatten)]` directly above the field with the comment above the hint
```suggestion
        /// Identifier to remove
        #[serde(flatten)]
```